\section{cluster\_dbscan\_X}

\subsection{General}

The \emph{cluster\_dbscan\_PCA}, \emph{cluster\_dbscan\_kmerL1},
\emph{cluster\_dbscan\_kmerL2}, \emph{cluster\_dbscan\_SW} and
\emph{cluster\_dbscan\_SW\_GPU} implement clustering of a dataset of
sequences using the DBSCAN \cite{dbscan} algorithm. The different
variants implement DBSCAN clustering using different sequence-sequence
distances and platforms. 

\subsection{Usage}

\lstset{language=bash,
  caption={Calling the \emph{cluster\_dbscan} tools},
  label=lst-dbscan-call}
\begin{lstlisting}
cluster_dbscan_pca [fasta] [pcaproj] [dimensions] [epsilon] [minPoints] \
   [outfiles-fasta-prefix] [outfiles-values-prefix]

cluster_dbscan_kmerL1 [fasta] [kmers] [epsilon] [minPoints] \
   [outfiles-fasta-prefix]

cluster_dbscan_kmerL2 [fasta] [kmers] [epsilon] [minPoints] \
   [outfiles-fasta-prefix]

cluster_dbscan_SW [fasta] [epsilon] [minPoints]
   [outfiles-fasta-prefix]

cluster_dbscan_SW_GPU [fasta] [epsilon] [minPoints] \
   [outfiles-fasta-prefix]   
\end{lstlisting}
Where the different tools perform as disguised by their name:
\begin{enumerate}
  \item \emph{cluster\_dbscan\_pca}: Clustering of sequences in a PCA
    subspace such as the one obtained using the \emph{kmer2pca} tool
    described in section \ref{sec-kmer2pca}.
  \item \emph{cluster\_dbscan\_kmerL1}: Clustering of sequences in a
    k-mer representation applying the $L_1$-norm also known as the
    Manhatten distance between the k-mer frequecy vectors that
    represent each sequence as a distance between them.
  \item \emph{cluster\_dbscan\_kmerL2}: Clustering of sequences in a
    k-mer representation applying the $L_2$-norm also known as the
    Manhatten distance between the k-mer frequecy vectors that
    represent each sequence as a distance between them.
  \item \emph{cluster\_dbscan\_SW}: Clustering of sequences using the
    Smith-Waterman distance measure.
  \item \emph{cluster\_dbscan\_SW\_GPU}: Clustering of sequences using
    the Smith-Waterman distance measure on an OpenCL supported device,
    for instance a GPU.
\end{enumerate}
The tools have the following arguments:
\begin{enumerate}
  \item \emph{fasta}: A FASTA file containing the sequences to be
    clustered. In the case of Smith-Waterman clustering the FASTA file
    has to containe nucleic sequences as clustering of proteic
    sequences under a Smith-Waterman distance is not implemented.
  \item \emph{pcaproj}: The projections onto the principal components
    as calculated by the \emph{kmer2pca} tool. 
  \item \emph{dimensions}: The number of dimensions, principal
    components found in the file.
  \item \emph{epsilon}: The epsilon neighbourhood radius as used by the DBSCAN
    \cite{dbscan} algorithm.
  \item \emph{minPoints}: The minimum number of sequences to be found
    within an epsilon neighbourhood in order to form or expand a
    cluster according to the DBSCAN algorithm.
  \item \emph{outfiles-fasta-prefix}: The DBSCAN based tools all
    generate a FASTA file per cluster found. With this argument you
    can tell the program where to store those clusters. The number of
    clusters to be generated is capped at 500 clusters. The resulting
    files will be named according to your prefix with a number at the
    end for each different cluster.
  \item \emph{outfiles-values-prefix}: If this parameter is added to
    the PCA version of the DBSCAN implementation the clusters are not
    only stored as individual fasta files but also as files containing
    the projections onto the principal components for the clustering
    in question. This for instance allows one to draw the different
    clusters into visual representations of the PCA projections.
\end{enumerate}

\subsection{Algorithm}

The utility implements the DBSCAN algorithm straight forward as
described in its original article \cite{dbscan}.

The different distance measures in the region expand method as
described in the original paper use various means of optimisation and
parallization.






    
